fhirr4/ballerina/src/main/resources/fhirservice/analytics_request_interceptor.bal

```diff
@@ -0,0 +1,143 @@
+configurable string[] requiredAttributes = ["fhirUser", "client_id", "iss"];
+
+// Configs to get more information about the patient
+configurable boolean moreInfoRequired = false;
+configurable string moreInfoUrl = "";
+configurable string moreInfoUsername = "";
+configurable string moreInfoPassword = "";
+
+const X_JWT_HEADER = "x-jwt-assertion";
+
+isolated service class AnalyticsRequestInterceptor {
+    *http:RequestInterceptor;
+    final string resourceType;
+
+    public function init(r4:ResourceAPIConfig apiConfig) {
+        self.resourceType = apiConfig.resourceType;
+    }
+
+    isolated function sendToOpenSearch(json logData) returns error? {
+        string url = opensearchUrl + "/" + opensearchIndex + "/_doc";
+        log:printDebug(`[AnalyticsRequestInterceptor] Sending log to: ${url} with index: ${opensearchIndex}`);
+
+        http:Client openSearchClient;
+        if opensearchUsername != "" && opensearchPassword != "" {
+            openSearchClient = check new (url, auth = {
+                username: opensearchUsername,
+                password: opensearchPassword
+            });
+        } else {
+            openSearchClient = check new (url);
+        }
+
+        http:Response result = check openSearchClient->/.post(logData);
+        if result.statusCode < 200 || result.statusCode >= 300 {
+            log:printError(`[AnalyticsRequestInterceptor] Failed publishing log to ${url} with Index: ${opensearchIndex}. ${"\n"}Error: ${result.reasonPhrase}`);
+        } else {
+            log:printInfo(`[AnalyticsRequestInterceptor] Log published successfully to ${url} with Index: ${opensearchIndex}`);
+        }
+    }
+
+    isolated function getMoreInfo(string patientId) returns json|error {
+        http:Client moreInfoClient;
+        if moreInfoUsername != "" && moreInfoPassword != "" {
+            moreInfoClient = check new (moreInfoUrl, auth = {
+                username: moreInfoUsername,
+                password: moreInfoPassword
+            });
+        } else {
+            moreInfoClient = check new (moreInfoUrl);
+        }
+        json moreInfoResponse = check moreInfoClient->/.get(patientId = patientId);
+        return moreInfoResponse;
+    }
+
+    isolated function publishAnalyticsData(string jwt) returns error? {
+        [jwt:Header, jwt:Payload]|error decodedJWT = jwt:decode(jwt);
+        if decodedJWT is [jwt:Header, jwt:Payload] {
+            [jwt:Header, jwt:Payload] [_, payload] = decodedJWT;
+
+            // Filter logData to include only keys in x_jwt_attributes
+            map<string> filteredLogData = map from string attrKey in requiredAttributes
+                where payload[attrKey] !== ()
+                select [attrKey, payload[attrKey].toString()];
+
+            filteredLogData["timestamp"] = time:utcToString(time:utcNow());
+            filteredLogData["resourceType"] = self.resourceType;
+
+            // Get more information of fhirUser
+            if moreInfoRequired {
```

Comment:
we can reduce nested if statements by following this: https://learn-ballerina.github.io/best_practices/early_returns.html